src/_posts/2020-03-22-types-and-accepting-the-facts.md

```diff
@@ -0,0 +1,525 @@
+
+Now let's take care of the IATA, the IATA specification (source: Wikipedia, I didn't read the specification) says that's a code composed by three letters. In this case, we can create a class for it and add the validation. 
+
+```java
+class InvalidIATAException extends InvalidArgumentException {}
+```
+
+```java
+class IATA {
+        
+    public final String iata;
+
+    public IATA(String iata) {
+        if (iata.length() != 3) {
+            throw new InvalidIATAException();
+        }
+        this.iata = iata;
+    }
+}
+```
+
+Then we change the service and the repository to start using types:
+
+```java
+class SearchService {
+
+    public List<Flight> findFlights(String startDate, String endDate, String origin, String dest) {
+        LocalDate flightStartDate = parseDate(startDate);
+        LocalDate flightEndDate = parseDate(endDate);
+        IATA originAirport = new IATA(origin);
+        IATA destAirport = new IATA(dest);
+        return searchRepository.findFlights(flightStartDate, flightEndDate, origin, dest);
+    }
+}
+```
+
+```java
+class SearchRepository {
+
+    public List<Flight> findFlights(LocalDate startDate, LocalDate endDate, IATA origin, IATA dest) {
+        // implementation
+    }
+}
+```    
+
+With those changes, we can at least guarantee that the dates passed to the database are valid and the formatting don't matter much at this point now because it's encapsulated inside of a class. We can't confirm that the date exists in the database but there won't be any exceptions thrown when the query is executed. 
+
+The state of the services and repositories are improving, said that there's still quite some work that can be done. We are doing all the validations inside the SearchService. 
+
+Just a clarification, The Application Layer is the part that handles the communication, in this case, it would be the controller. The controller isn't related to the business but just a way of input/output for our domain. This diagram shows the boundaries between them:
+
+![Diagram of the application with the controller, flight service and flight repository]({{ '/assets/custom/img/blog/2020-03-22-types-and-accepting-the-facts/application-diagram.jpg' | prepend: site.baseurl }})
+
+We use types, dependency injection and interfaces to abstract what the application is doing. The `FlightSearchService` don't care if the data is coming through HTTP, RPC or even a CLI. The same goes for the `FlightSearchRepository`, it just cares that you can store and retrieve the data later, the how doesn't matter for the business, that's an application responsibility. 
+
+Continuing with the changes. Now we can see that the problem with the codebase is that is the service which is part of our business domain and we are filling with application code that has nothing to do with it. The solution for that is moving that code up the application layer.
```

Comment:
Change this sentence: Now we can see that the problem with the codebase is that is the service which is part of our business domain and we are filling with application code that has nothing to do with it.

Suggested sentence: The problem with the codebase is that it's being filled with application code that isn't relevant.